src/sentry/db/models/paranoia.py

```diff
@@ -0,0 +1,39 @@
+from __future__ import absolute_import
+
+from django.db import models
+from django.db.models import Manager
+from django.db.models.query import QuerySet
+from django.utils import timezone
+
+from sentry.db.models import Model
+
+
+class ParanoidQuerySet(QuerySet):
+    """
+    Prevents objects from being hard-deleted. Instead, sets the
+    ``date_deleted``, effectively soft-deleting the object.
+    """
+
+    def delete(self):
+        self.update(date_deleted=timezone.now())
```

Comment:
I'm ok with this, but I'll just want to point out that if you end up doing any UNIQUE constraints or anything like that, we'll need to specially craft an index to handle this. Since with a soft delete, you may still have a row that violates your UNIQUEness.

This is usually a bit more aware when it's not abstracted out into a subclass like this hiding the behavior.

Not anything to change, but something to keep in mind.

An example of this would be say, a slug. If you soft delete, but slug is UNIQUE, that means slugs will never be able to be reused unless you also override the delete behavior to change the slugs to NULL or some other value to avoid the dupe. Or we create a functional index (that would only work on Postgres).

Reply:
I think it's fine to have `slug` be unique (it probably should be, now that I think about it) but _not_ worry about freeing up that name on soft-delete. Once a `SentryApp` is created, I don't think we'd want another service to be able to, effectively, imitate it if it were ever removed removed.

Reply:
Yup, that's fine then. Just wanted to make sure the intention will always be to undelete to reuse the old row, rather than trying to create new ones, etc.